webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/RuleApi.scala

```diff
@@ -933,25 +935,52 @@ class RuleApiService14 (
     }
   }
 
+  private def listCategoriesId(cat: RuleCategory) : List[RuleCategoryId] = {
+    @tailrec
+    def listCatAcc(categories: List[RuleCategory], acc: List[RuleCategoryId]) : List[RuleCategoryId] = {
+      categories match {
+        case Nil => acc
+        case c :: t => listCatAcc(t, c.id :: acc)
+      }
+    }
+    listCatAcc(cat.childs, List(RuleCategoryId("rootRuleCategory")))
+  }
+
+  // List all categories mentioned in categoryId in rules who are not in database
+  def getMissingCategories(existingCat: RuleCategory, rules: List[Rule]): List[RuleCategory] = {
+    val catIds = listCategoriesId(existingCat)
+    val rulesCat = rules.map(_.categoryId.value)
```

Comment:
You should keep to RuleCategoryId object and not their value (you can compare directly case classes, you don't need to go to the value)

Moreover here you can you function diff of a list that contains onlyuelement of a list that are not in the second list

so you should 

rulesCat = rules.map(_.categoryId)
rulesCat.diff(catIds)

Reply:
you should do the distinct also on rulesCat